# Цели исследования: 

Проанализировать таблицы и ответить на следующие вопросы:

1) Посчитайте, сколько книг вышло после 1 января 2000 года;

2) Для каждой книги посчитайте количество обзоров и среднюю оценку;

3) Определите издательство, которое выпустило наибольшее число книг толще
50 страниц — так вы исключите из анализа брошюры;

4) Определите автора с самой высокой средней оценкой книг — учитывайте
только книги с 50 и более оценками;

5) Посчитайте среднее количество обзоров от пользователей, которые поставили
больше 48 оценок.

In [94]:
#Подключение библиотек

import pandas as pd
from sqlalchemy import text, create_engine

In [95]:
db_config = {'user': 'praktikum_student', # имя пользователя
    'pwd': 'Sdf4$2;d-d30pp', # пароль
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432, # порт подключения
    'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [96]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Открытие таблицы books

In [97]:
# чтобы выполнить SQL-запрос, используем Pandas
query_books = '''SELECT * FROM books'''

In [98]:
con = engine.connect()

In [99]:
pd.io.sql.read_sql(sql=text(query_books), con = con)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

# Открытие таблицы authors

In [100]:
# чтобы выполнить SQL-запрос, используем Pandas
query_authors = '''SELECT * FROM authors LIMIT 5'''

In [101]:
pd.io.sql.read_sql(sql=text(query_authors), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

# Открытие таблицы publishers

In [102]:
# чтобы выполнить SQL-запрос, используем Pandas
query_publishers = '''SELECT * FROM publishers LIMIT 5'''

In [103]:
pd.io.sql.read_sql(sql=text(query_publishers), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

# Открытие таблицы ratings

In [104]:
# чтобы выполнить SQL-запрос, используем Pandas
query_ratings = '''SELECT * FROM ratings LIMIT 5'''

In [105]:
pd.io.sql.read_sql(sql=text(query_ratings), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

# Открытие таблицы reviews

In [106]:
# чтобы выполнить SQL-запрос, используем Pandas
query_reviews = '''SELECT * FROM reviews LIMIT 5'''

In [107]:
pd.io.sql.read_sql(sql=text(query_reviews), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

# Первый запрос

## В первом запросе требуется посчитать сколько книг вышло после 01.01.2000 года

In [108]:
quest_one = '''SELECT COUNT(book_id) 
            FROM books
            WHERE publication_date >= '2000-01-01' '''

In [109]:
pd.io.sql.read_sql(sql=text(quest_one), con = con)

count
0    821

821 книга вышла после 2000 года

# Второй запрос

## Требуется посчитать количество обзоров и среднюю оценку для каждой книги

In [201]:
quest_two = '''SELECT 
                b.book_id,
                b.title,
                COUNT(DISTINCT r.review_id) AS count_of_reviews,
                AVG(rat.rating) AS mean_rating
            FROM books AS b
            LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
            LEFT JOIN reviews AS r ON b.book_id = r.book_id
            GROUP BY b.book_id'''

In [202]:
pd.io.sql.read_sql(sql=text(quest_two), con = con)

book_id                                              title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997                        Xenocide (Ender's Saga  #3)   
997      998                                    Year of Wonders   
998      999                        You Suck (A Love Story  #2)   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...   

     count_of_reviews  mean_rating  
0                   2     3.666667  
1                   1     2.500000  
2                   3     4.666667  
3                   2     4.500000  
4                   4     4.000000  
..                ...          ...  
995                 3     3.666667  
996                 3     3.400000  
997                 4     3.200000  
998                 2     4.500000  
999                 4     3.833333  

[1000 rows x 4 columns]

# Третий запрос

## Требуется определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [112]:
quest_three = '''SELECT p.publisher,
                COUNT(b.book_id) AS count_of_books
            FROM books AS b
            INNER JOIN publishers AS p ON b.publisher_id = p.publisher_id
            WHERE b.num_pages > 50
            GROUP BY publisher
            ORDER BY count_of_books DESC
            LIMIT 1'''

In [113]:
pd.io.sql.read_sql(sql=text(quest_three), con = con)

publisher  count_of_books
0  Penguin Books              42

Penguin Books - издательство, которое выпустило наибольшее число книг толще 50 страниц.

# Четвёртый запрос

## Нужно определить автора с самой высокой средней оценкой книг, у которых более 50 оценок

In [203]:
quest_four = '''SELECT a.author,
                COUNT(rat.rating) AS c_ratings,
                AVG(rat.rating) AS avg
            FROM authors AS a
            INNER JOIN books AS b ON a.author_id = b.author_id
            INNER JOIN ratings AS rat ON rat.book_id = b.book_id
            WHERE b.book_id IN
                (SELECT b.book_id
                 FROM books AS b
                 INNER JOIN ratings AS rat ON b.book_id = rat.book_id
                 GROUP BY b.book_id
                 HAVING COUNT(rat.rating_id) >= 50)
            GROUP BY author
            ORDER BY c_ratings DESC
            LIMIT 1'''

In [204]:
pd.io.sql.read_sql(sql=text(quest_four), con = con)

author  c_ratings       avg
0  J.K. Rowling/Mary GrandPré        310  4.287097

J.K. Rowling/Mary GrandPré - автора с самой высокой средней оценкой книг, у которых более 50 оценок.

# Пятый запрос

## Требуется посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [171]:
quest_five = '''WITH a as
    (SELECT username,
    COUNT(review_id) as cri
     FROM reviews
     WHERE username in 
                    (SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating) > 48)
                    GROUP BY username)
    
    SELECT AVG(cri)
    FROM a'''

In [172]:
pd.io.sql.read_sql(sql=text(quest_five), con = con)

avg
0  24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24

# Общий вывод: в ходе исследования были проанализированы 5 таблиц, среди которых:
1) books - содержит инфрормацию о книгах;

2) authors - содержит данные об авторах;

3) publishers - содержит данные об издательствах;

4) ratings - содержит данные о пользовательских оценках книг;

5) reviews - содержит данные о пользовательских обзорах на книги.

В процессе исследования было выявлено:
    после начала 2000 года вышла 821 книга;
    для каждой книги нашли среднюю оценку и количество обзоров;
    Penguin Books - издательство, которое выпустило наибольшее число книг толще 50 страниц - 42;
    автор с самой высокой средней оценкой книг, у которых более 50 оценок - J.K. Rowling/Mary GrandPré;
    среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.0